### ISSUE TITLE
fallback/progressive enhancement

### ISSUE THREAD

**Issue by thepassle** [thepassle/generic-components#10]
Components this will probably be most relevant to:
- generic-accordion
- generic-tabs
- generic-disclosure

Currently, its recommended to use a button for the 'headers' or 'controls' of an accordion/tabs component. The problem with this is progressive enhancement, or a fallback if JS is not available; in that case, there will be a button, but that button does not really do anything or 'control' anything, which can be confusing to a user.

It may make more sense to just simply fall back to some kind of heading or text node (whatever the user should provide) instead. 

If the goal of this project is to 'enhance' markup with some functionality like keyboard behavior and accessibility, that might be a nice/elegant fallback.

That might also mean it may make more sense to 'enforce' or recommend a pattern like, which is similar to a `detail` & `summary` pattern:

```html
<generic-accordion>
  <generic-accordion-header>Shipping information</generic-accordion-header>
  <generic-accordion-panel>Here is some shipping information</generic-accordion-panel>
  
  <generic-accordion-header>Personal information</generic-accordion-header>
  <generic-accordion-panel>Here is some personal information</generic-accordion-panel>
</generic-accordion>
```

**Comment by bkardell** [comment]
I would avoid option 1.. we kind of considered going this way in our original proposal but the relationships are really not great, it's really as if those were just 1 element of text.

Any of the other 3 might have good arguments for or against.  The third one seems interesting but I'm not sure it's actually practically useful & might have subtle challenges of wanting to move the focus navigation start point, change the url hash of the browser unwantedly and initiate scroll behavior. The second or third would be my guess as something that strikes pretty good balance. Not really sure what you mean about html not having attributes, there's plenty of them, Boolean ones too, but even if not the same could be said about slots being exposed, and I think it's ok because we're consciously trying something new here.  My only slight hesitation with slots is that if you also work in some kind of dynamic projection to accommodate both layouts it would have to override this, but I guess the same is true about if you explicitly wrote some Aria on those so maybe it's ok.. internals and aom will make this slightly better I think in terms of not having to sprout new attributes in your light Dom (I hope)

**Comment by thepassle** [comment]
Another option could be:

## 5: `header-for` and `idref`

pros: 
- similar syntax to `aria-labelledby` or `aria-describedby`
- requires consumer to be explicit about relations between a header and a panel
- allows the user to be able to truly compose the tabs however they want (e.g.: tabs on the side, or bottom)

cons:
- lose any possibility to have default style with `:host` (e.g.: a default `display` property)
  - example: by default it would show the headers underneath either, not next to eachother like a 'tab-style'.

```html
<generic-tabs> 
  <h2 header-for="panel-1">Shipping information</h2>
  <h2 header-for="panel-2">Personal information</h2>
  <div id="panel-1">
    Here is some shipping information
  </div>
  <div id="panel-2">
    Here is some personal information
  </div>
</generic-tabs>
```

**Comment by kurtrank** [comment]
I agree, I think for most components falling back to some sort of text node is a good plan. I have experimented with building some web components and I was leaning towards the same way.

## Accordion

According to the WAI-ARIA example, accordions are used for sections of content and should have heading elements of an appropriate level. I think option 3 works perfect here. As far as the styling goes, I think that is up to the page that is up to the author. Some sites may want to inherit the different heading level styles or some may want all heading levels to look the same in accordions, but both could be possible.

``` css
/* style all heading tags used as accordion headers */
generic-accordion > [slot="header"] {
  font-size: 1em;
  margin: 0;
}
```

## Tabs

Tabs still manage multiple sections of content, but instead of the headings being the primary source of navigation there are a list of controls at the top, essentially like a table of contents. I think option 4 makes the most sense here. The tabs would start as anchor tags grouped at the top so that they can jump to the section (like in Zach's seven minute tabs), but then be converted into buttons.

``` html
<generic-tabs>
  <a slot="tab" href="#tab-1">Shipping Information</a>
  <a slot="tab" href="#tab-2">Personal Information</a>
  <div slot="panel" id="tab-1">
    Here is some shipping information
  </div>
  <div slot="panel" id="tab-2">
    Here is some personal information
  </div>
</generic-tab>
```

So pretty close to your current implementation but just starting as anchor elements with matching IDs. The big downside I can see to this is that authors have to manually add IDs. I like not having to worry about that but otherwise there's no point in starting the tab controls as anchors.

## Disclosure

I would say this matches the tabs scenario, where it should start out as something else (paragraph, span, etc) and then is converted into a button when the component is initialized.

---

It seems to me that each component might fit best into different solutions that you proposed. Are you planning on trying to come up with a more consistent approach that would be used for all of them?

**Comment by thepassle** [comment]
> It seems to me that each component might fit best into different solutions that you proposed. Are you planning on trying to come up with a more consistent approach that would be used for all of them?

It would be nice to have some kind of strategy for every component eventually yeah. But I think it'd have to be on a component to component basis, see what the best solution is for each component, and then potentially align across the board where possible. So thanks for your input! Very valuable 🙂

My current thinking is that just using slots is probably the best way forward to both be able to support a fallback, as well as a nice developer experience.

**Comment by bkardell** [comment]
I mean, I am biased but I think that #2 in your list above is best (basically https://panel-set.glitch.me/ is just this with some attribute sugar).  The thing I like about this approach is that it's _just_ HTML, there's no kind of in-between "half way" where you have semi-nonsensical HTML nobody would probably write if there's no upgrade....  It doesn't pretend, you know?  It's legit, sensible, non-interactive HTML that upgrades to interactive tabs (or vertical tabs/accordion-style).

**Comment by Malvoz** [comment]
https://github.com/thepassle/generic-components/issues/10#issue-611523936:
> Currently, its 'recommended' to use a button for the 'headers' or 'controls' of an accordion/tabs component. The problem with this is progressive enhancement, or a fallback if JS is not available; in that case, there will be a button

Due to `:not(:defined) { display: none }`, _nothing_ will actually be displayed to the user in browsers that support custom elements (and subsequently the `:defined` pseudo class) when JS is disabled.

My approach in solving this (for the Maps for HTML Community Group's Web-Map-Custom-Element) was to include to [include a `<noscript>` element and reset `display: none`](https://github.com/Maps4HTML/Web-Map-Custom-Element/blob/01f6bd5f866acefbea4f564bb576a271b27b6348/index-mapml-viewer.html#L46-L54). FWIW this could be done prettier if browsers decided to implement the [scripting `@media` feature](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/scripting).
